chain/sync.go

```diff
@@ -522,7 +526,8 @@ func (syncer *Syncer) ValidateBlock(ctx context.Context, b *types.FullBlock) err
 		log.Warn("timestamp funtimes: ", h.Timestamp, baseTs.MinTimestamp(), h.Height, baseTs.Height())
 		diff := (baseTs.MinTimestamp() + (build.BlockDelay * uint64(h.Height-baseTs.Height()))) - h.Timestamp
 
-		return xerrors.Errorf("block was generated too soon (h.ts:%d < base.mints:%d + BLOCK_DELAY:%d * deltaH:%d; diff %d)", h.Timestamp, baseTs.MinTimestamp(), build.BlockDelay, h.Height-baseTs.Height(), diff)
+		return validation.ErrorWrapString(validation.ErrBlockFutureTimestamp, fmt.Sprintf("h.ts:%d < base.mints:%d + BLOCK_DELAY:%d * deltaH:%d; diff %d",
```

Comment:
So I really like the intention of this PR, my only issue is that the `ErrorWrapString` method prevents the use of structured logging for the values in the message, I would love for the message to be formatted so the log line comes out as follows:
```
{"level": "error", "msg": "h.ts < base.mints + BLOCK_DELAY * deltaH", "ts": <ts value>,"base.mints": <base.mints value>, "block.delay": <block delay value>, "height.delta": <height delta value>, "diff": <diff value>}
```

Reply:
Fair enough, but that is outside the scope of this effort.